### ISSUE TITLE
improve SMART parameters standardization (ex 'full') 

### ISSUE THREAD

**Issue by DominiqueMakowski** [easystats/parameters#97]
imo this is a critical issue to be able to retrieve 'refit' standardized parameters with a posthoc method.

improvements are possible with especially for the case of interaction, but a more robust and systematic testing framework might be needed. also, a knowledge of model matrices when factors are involved appears as key.

**Comment by strengejacke** [comment]
❓ 

I have no idea what you're talking about 😆  I _could_ look at the code, but am too lazy... Can you elaborate a bit more, maybe with an example?

**Comment by DominiqueMakowski** [comment]
Long story short they are still [some cases](https://easystats.github.io/parameters/articles/standardization.html#interactions-1) where the SMART method does not perform well, in comparison to refit and also to "classic". This appears mainly for interaction terms, and for interaction between a continuous and a factor variable. 

I have no idea how to improve on that, and I think that to address this we would need someone with a deep understanding on how model matrices are built for interactions terms, and how we can standardize in it the interaction term so that it reflects the interaction term of two standardized variables...

That's more a longterm issue tho, if by any chance we meet someone with some understanding of the model matrices and formulas...

**Comment by DominiqueMakowski** [comment]
In a nutshell, the goal here is to reconstruct the standardized model.matrix (the 2nd one below) with the original model matrix and the Mean/SD of each variable...

``` r
df <- iris
dfZ <- parameters::standardize(iris)

head(model.matrix(~ df$Sepal.Length * df$Species))
#>   (Intercept) df$Sepal.Length df$Speciesversicolor df$Speciesvirginica
#> 1           1             5.1                    0                   0
#> 2           1             4.9                    0                   0
#> 3           1             4.7                    0                   0
#> 4           1             4.6                    0                   0
#> 5           1             5.0                    0                   0
#> 6           1             5.4                    0                   0
#>   df$Sepal.Length:df$Speciesversicolor df$Sepal.Length:df$Speciesvirginica
#> 1                                    0                                   0
#> 2                                    0                                   0
#> 3                                    0                                   0
#> 4                                    0                                   0
#> 5                                    0                                   0
#> 6                                    0                                   0
head(model.matrix(~ dfZ$Sepal.Length * dfZ$Species))
#>   (Intercept) dfZ$Sepal.Length dfZ$Speciesversicolor dfZ$Speciesvirginica
#> 1           1       -0.8976739                     0                    0
#> 2           1       -1.1392005                     0                    0
#> 3           1       -1.3807271                     0                    0
#> 4           1       -1.5014904                     0                    0
#> 5           1       -1.0184372                     0                    0
#> 6           1       -0.5353840                     0                    0
#>   dfZ$Sepal.Length:dfZ$Speciesversicolor
#> 1                                      0
#> 2                                      0
#> 3                                      0
#> 4                                      0
#> 5                                      0
#> 6                                      0
#>   dfZ$Sepal.Length:dfZ$Speciesvirginica
#> 1                                     0
#> 2                                     0
#> 3                                     0
#> 4                                     0
#> 5                                     0
#> 6                                     0
```

<sup>Created on 2019-09-23 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by DominiqueMakowski** [comment]
I believe one of the reasons for the issues with interactions stems out of the fact that as we know, a regression model "fixes" the other parameters at 0. This corresponds to the mean, when a standardized dataset is passed.

Let's say we have the interaction between x * y. The coefficient corresponding to "x" is the coefficient of x when y = 0. This coefficient changes, as y changes (following the interaction coefficient). Now, if a standardized dataset is passed, it is normal that the effect of the "x" parameter is different, as it corresponds to the effect of "x" at the mean of "y" (which might not be the case of unstandardized data).

Hence, my hint is that the posthoc standardization should somehow take the mean of the variables into account in the case of interactions. I have no idea how, though.

To facilitate the exploration, I've refacted parameters_standardize and created the `standardize_info` function, that returns values useful for parameters standardization, such as deviations of response and variables.

``` r
model <- lm(Sepal.Width ~ Petal.Width * Sepal.Length, data = iris)
info <- parameters::standardize_info(model)
info$Refit <- parameters::parameters_standardize(model, method = "refit")[, 2]
info$Raw <- insight::get_parameters(model)[, 2]
info[sapply(info, is.numeric)] <- sapply(info[sapply(info, is.numeric)], round, digits = 1)
info
#>                  Parameter        Type Factor Deviation_Response
#> 1              (Intercept)   intercept   <NA>                0.4
#> 2              Petal.Width     numeric  FALSE                0.4
#> 3             Sepal.Length     numeric  FALSE                0.4
#> 4 Petal.Width:Sepal.Length interaction  FALSE                0.4
#>   Mean_Response Deviation_Classic Mean_Classic Deviation_Smart Mean_Smart
#> 1           3.1               0.0          0.0             0.0        0.0
#> 2           3.1               0.8          1.2             0.8        1.2
#> 3           3.1               0.8          5.8             0.8        5.8
#> 4           3.1               5.3          7.5             0.8        5.8
#>   Refit  Raw
#> 1  -0.2  3.4
#> 2  -0.7 -1.5
#> 3   0.4  0.0
#> 4   0.3  0.2
```

<sup>Created on 2019-10-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

In a nutshell, the problem is to try to FIND the "Refit" column from the "Raw" column using the remaining information...

**Comment by DominiqueMakowski** [comment]
At the same time, it suggests that the issue with interaction are not real issues, it's just that the estimates correspond to something different, but they are not wrong per se (*I think*)

**Comment by DominiqueMakowski** [comment]
I reckon that's the reason for partial standardized coefficients (https://www.jstor.org/stable/2684719), but we need VIF for that
